### PR TITLE
[mlir][python] remove various caching mechanisms

### PR DESCRIPTION
**Pull request by makslevental**
This PR removes the various caching mechanisms currently in the python bindings - both positive caching and negative caching.

### CODE REVIEW

File: mlir/lib/Bindings/Python/MainModule.cpp [hunk 1]
```diff
@@ -34,7 +34,6 @@ PYBIND11_MODULE(_mlir, m) {
           "append_dialect_search_prefix",
           [](PyGlobals &self, std::string moduleName) {
             self.getDialectSearchPrefixes().push_back(std::move(moduleName));
-            self.clearImportCache();
```
Comment (makslevental) [inline mlir/lib/Bindings/Python/MainModule.cpp:37]: The only question is whether there's any value in preserving these semantics for the  - i.e., that when you append a dialect search prefix, that `llvm::StringSet<> loadedDialectModules` should be cleared thereby triggering "reloads" of existing dialect modules. 

I put "reload" in quotes because according to my understanding a reimport doesn't actually trigger a reload of the module[^1]. So even if someone is dynamically moving around existing modules (expecting dialect modules found under former prefixes to be found elsewhere post a call to `append_dialect_search_prefix`), the current implementation doesn't actually do that (since it reimports by calling `py::module::import(moduleName.c_str())`).

[^1]: You need to use `importlib` [to accomplish this](https://stackoverflow.com/questions/437589/how-do-i-unload-reload-a-python-module/437591#437591).

Comment (rkayaith) [reply mlir/lib/Bindings/Python/MainModule.cpp:37]: My interpretation is that this is to support:
1. `loadDialectModule("my_dialect")` - doesn't find module
2. `append_dialect_search_prefix("path/to/my_dialect")`
3. `loadDialectModule("my_dialect")` - finds module
 
So its not about reloading existing modules, just loading new ones. Without clearing `loadedDialectModules` you'd end up hitting the early exit in `loadDialectModule`.

File: mlir/lib/Bindings/Python/IRModule.cpp [hunk 2]
```diff
@@ -108,86 +105,51 @@ void PyGlobals::registerOperationImpl(const std::string &operationName,
...
-  // Not found and loading did not yield a registration. Negative cache.
-  dialectClassMap[dialectNamespace] = py::none();
+  // Not found and loading did not yield a registration.
   return std::nullopt;
 }
 
 std::optional<pybind11::object>
 PyGlobals::lookupOperationClass(llvm::StringRef operationName) {
+  // Fast match against the class map first (succeeds if second lookup, after
+  // successful dialect load).
```
Comment (rkayaith) [inline mlir/lib/Bindings/Python/IRModule.cpp:148]: any particular reason this one does a double-lookup, but the others `loadDialectModule` first and only do one? in the absence of perf data, might be better to lean towards making them all consistent

**Review by rkayaith — COMMENTED** [review]
nice, I never fully understood all those caches

File: mlir/lib/Bindings/Python/MainModule.cpp (hunk 1 above)
Comment (makslevental) [reply mlir/lib/Bindings/Python/MainModule.cpp:37]: Ah right that makes sense. Cool I'll restore the "clear" then.

Comment (makslevental) [reply mlir/lib/Bindings/Python/MainModule.cpp:37]: On second thought, let's make `loadDialectModule` do the right thing - report whether a dialect is actually registered or not (instead of just "pretending" a dialect is loaded).

File: mlir/lib/Bindings/Python/IRModule.cpp [hunk 3]
```diff
@@ -57,15 +57,18 @@ void PyGlobals::loadDialectModule(llvm::StringRef dialectNamespace) {
     break;
   }
 
+  if (loaded.is_none())
+    return false;
   // Note: Iterator cannot be shared from prior to loading, since re-entrancy
   // may have occurred, which may do anything.
-  loadedDialectModulesCache.insert(dialectNamespace);
+  loadedDialectModules.insert(dialectNamespace);
+  return true;
```
Comment (rkayaith) [inline mlir/lib/Bindings/Python/IRModule.cpp:65]: I think the negative caching that used to happen here actually made sense (avoiding python interpreter work seems significant vs just C++ map lookups that the other caching potentially avoided)

Comment (ftynse) [reply mlir/lib/Bindings/Python/IRModule.cpp:65]: I don't see any previous negative caching at module level, am I missing something?

It could make sense to add _here_. Several lookup functions are calling this, and in absence of the module for a specific dialect prefix, this will keep hitting the slow path of trying to load a module from a filesystem. This seems undesirable unless module import has its own cache. (I generally prefer adding optimizations when we know they are actually meaningful...) We should document this behavior and provide a way to reset the (negative) cache programmatically for extreme cases like the module being placed into one of the search paths by another process.

File: mlir/python/mlir/ir.py [hunk 4]
```diff
@@ -277,15 +277,15 @@ def _f64ElementsAttr(x, context):
     def _i32ElementsAttr(x, context):
         return DenseElementsAttr.get(
             np.array(x, dtype=np.int32),
-            type=IntegerType.get_signed(32, context=context),
+            type=IntegerType.get_signless(32, context=context),
```
Comment (ftynse) [inline mlir/python/mlir/ir.py:280]: This change looks irrelevant to the patch

File: mlir/test/python/ir/custom_dialect/custom.py
Comment (ftynse) [inline mlir/test/python/ir/custom_dialect/custom.py:1]: Could we have a comment that describes the purpose of this file (for testing)?

File: mlir/test/python/ir/insertion_point.py [hunk 5]
```diff
@@ -3,7 +3,13 @@
 import gc
 import io
 import itertools
+import sys
+
 from mlir.ir import *
+from mlir.dialects._ods_common import _cext
+
+sys.path.append(".")
+_cext.globals.append_dialect_search_prefix("custom_dialect")
```
Comment (ftynse) [inline mlir/test/python/ir/insertion_point.py:12]: I suppose this piggy-backs on the test logic below that creates an operation with "custom" dialect namespace, that is supposed to somehow trigger the dialect loading? It's not obvious from the test what would happen if loading doesn't happen, and how to fix it, so I'd much rather have a separate test for the loading mechanism itself. We can add `_private` methods to `PyGlobals` class for testing purposes, there is precedent. Such a method can, for example, tell you if a dialect prefix is present in the set of loaded dialects.

File: mlir/lib/Bindings/Python/IRModule.cpp (hunk 3 above)
Comment (makslevental) [reply mlir/lib/Bindings/Python/IRModule.cpp:65]: > I don't see any previous negative caching at module level, am I missing something?

The negative cache currently works in that when a dialect is "looked up" but fails to have its module found (for lack of existence or lack of prefix search path) will still be `loadedDialectModulesCache.insert(dialectNamespace);` (the loop just finishes silently).

> Several lookup functions are calling this

That's true but there's only one (under reasonable circumstances) that could possibly actually trigger a load that doesn't succeed (as far as I can imagine): `lookupOperationClass` during `Operation.create` and only if you're creating an operation with/from an unregistered dialect.

> absence of the module for a specific dialect prefix

This is what the added test code below exercises, demonstrates (although obliquely). In the absense of the added `custom_dialect.custom.py`, it's exactly as you say (the lookup happens over and over) but with the addition of that module it only happens once.

File: mlir/test/python/ir/insertion_point.py (hunk 5 above)
Comment (makslevental) [reply mlir/test/python/ir/insertion_point.py:12]: > It's not obvious from the test what would happen if loading doesn't happen

Yes I couldn't figure out how to "surface" the "failing" behavior discussed above - that without this `custom_dialect.custom.py`, `loadDialectModule` happens over and over. 

> Such a method can, for example, tell you if a dialect prefix is present in the set of loaded dialects.

Okay I'll do that.

File: mlir/lib/Bindings/Python/IRModule.cpp (hunk 3 above)
Comment (ftynse) [reply mlir/lib/Bindings/Python/IRModule.cpp:65]: Thanks for the explanation!

> That's true but there's only one (under reasonable circumstances) that could possibly actually trigger a load that doesn't succeed (as far as I can imagine)

If we have a type/attr from an unregisterd/unloadable dialect, I believe we can still process that opaquely and have a python object representing that type/attr. Wouldn't obtaining such an object trigger caster lookup that always lead to repeated module searches?

> In the absense of the added custom_dialect.custom.py, it's exactly as you say (the lookup happens over and over) but with the addition of that module it only happens once.

I suppose creating ops from an unloaded/unregistered dialect isn't very common, so it's not critical to improve that, but still nice to have.

Comment (makslevental) [reply mlir/lib/Bindings/Python/IRModule.cpp:65]: Yea so I guess it's true you can get into these weird corners where you want to do stuff in python but you haven't setup any of the boiler plate? I agree that I don't think it's common and adding "cache invalidation" mechanisms, such as we have currently (and this PR is meant to excise) to `loadedDialectModules` probably isn't worth it.

Comment (jpienaar) [reply mlir/lib/Bindings/Python/IRModule.cpp:65]: Is there any follow up needed here? E.g., adding a TODO or github issue to track? (It sounds like agreement that this is only bad in what should be a very uncommon case and the added complexity/footgun doesn't warrant it).

**Review by jpienaar — APPROVED** [review]
Looks like agreement that this cache isn't doing much here. So LGTM to remove and could be revised if we see regression/get reports.
